### PR TITLE
Proof serialization

### PR DESCRIPTION
**Pull request by rjeli**
Mostly straightforward, two iffy parts:
- ok to put `array_serialization` in `p3-util` (pulling in `serde`), or should be its own crate?
- ~Had to wrap Merkle commitment to be serializable, and then to satisfy `CanObserve`, changed it so `Commitments` are `IntoIterator` with `Item` observable. kind of ugly. but not terrible.~

### CODE REVIEW

**Comment by rjeli** [comment]
Alternatives:
1. put `[P::Scalar; DIGEST_ELEMS]: Serialize + for<'de> Deserialize<'de>` in `FieldMerkleTreeMmcs`, `FriConfig`, and `StarkConfig`
2. move `FieldDigest` to `p3-symmetric` or `p3-challenger`

1 is probably best imo, lmk and I will change

**Comment by rjeli** [comment]
I changed it to 1 so it's not ugly.

File: util/src/array_serialization.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    fn expecting(&self, formatter: &mut core::fmt::Formatter) -> core::fmt::Result {
+        formatter.write_fmt(format_args!("an array of length {}", N))
+    }
+
+    #[inline]
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: SeqAccess<'de>,
+    {
+        let mut data = Vec::with_capacity(N);
```
Comment (dlubarov) [inline util/src/array_serialization.rs:37]: nit: can we use an array upfront rather than doing `try_into()` later?

**Comment by rjeli** [comment]
another alternative is to remove all serde bounds from `FriConfig`, `StarkConfig`, and put them in `#[serde(bound)]` to "defer" it to the downstream user trying to serialize. that way is "more correct" but honestly it doesn't seem hard to implement `Commitment: Serde` for all PCS, and means serde bounds spam in data structures

**Review by dlubarov — APPROVED** [review]
LGTM

File: util/src/array_serialization.rs (hunk 1 above)
Comment (rjeli) [reply util/src/array_serialization.rs:37]: with `MaybeUninit`?

**Pull request merged**